### PR TITLE
With htslib

### PR DESCRIPTION
**Pull request by DrYak**
Fixing issue #47 

Branch actually *NOT* finished :
 - Meson needs to be adapted too

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -28,52 +26,15 @@ diri_sampler_LDADD = $(GSL_LIBS)
 b2w_SOURCES = \
 	src/cpp/b2w.cpp
 b2w_CXXFLAGS = $(PTHREAD_CFLAGS)
-b2w_CPPFLAGS = -I$(top_srcdir)/src/samtools
-b2w_LDADD = libbam.a $(ZLIB_LIBS) $(PTHREAD_LIBS)
+b2w_CPPFLAGS = $(HTSLIB_CPPFLAGS)
```
Comment (SoapZA) [inline Makefile.am]: `HTSLIB_CPPFLAGS` doesn't exist, `HTSLIB_CFLAGS` is the flag you're looking for

File: Makefile.am [hunk 2]
```diff
@@ -28,52 +26,15 @@ diri_sampler_LDADD = $(GSL_LIBS)
...
+b2w_LDADD = $(HTSLIB_LDFLAGS) -lhts  $(ZLIB_LIBS) $(PTHREAD_LIBS)
```
Comment (SoapZA) [inline Makefile.am]: same here, `HTSLIB_LDFLAGS` -> `HTSLIB_LIBS`

File: Makefile.am [hunk 3]
```diff
@@ -28,52 +26,15 @@ diri_sampler_LDADD = $(GSL_LIBS)
...
 
 fil_SOURCES = \
 	src/cpp/fil.cpp
 fil_CXXFLAGS = $(PTHREAD_CFLAGS)
-fil_CPPFLAGS = $(GSL_CFLAGS) -I$(top_srcdir)/src/samtools
-fil_LDADD = libbam.a $(ZLIB_LIBS) $(GSL_LIBS) $(PTHREAD_LIBS)
+fil_CPPFLAGS = $(GSL_CFLAGS) $(HTSLIB_CPPFLAGS)
+fil_LDADD = $(HTSLIB_LDFLAGS) -lhts $(ZLIB_LIBS) $(GSL_LIBS) $(PTHREAD_LIBS)
```
Comment (SoapZA) [inline Makefile.am]: ditto, and remove `-lhts`

File: configure.ac [hunk 4]
```diff
@@ -88,7 +88,7 @@ AS_IF([test "x${enable_flags_setting}" = "xyes" && test "x${enable_debug}" = "xn
 	CFLAGS=$( echo ${CFLAGS} | $SED -e 's/^ *//' -e 's/ *$//' )
 
 	AC_LANG([C++])
-	AX_APPEND_COMPILE_FLAGS([${cflags_test}], [CXXFLAGS])
+	AX_APPEND_COMPILE_FLAGS([${cflags_test} -std=gnu++11], [CXXFLAGS])
```
Comment (SoapZA) [inline configure.ac]: Wrong macro, please use https://www.gnu.org/software/autoconf-archive/ax_cxx_compile_stdcxx.html#ax_cxx_compile_stdcxx
`AX_CXX_COMPILE_STDCXX([11], [noext], [mandatory])`

File: configure.ac [hunk 5]
```diff
@@ -132,6 +132,21 @@ PKG_CHECK_MODULES([GSL], [gsl])
...
+dnl ===========
+AX_WITH_HTSLIB
+if test "$ax_cv_htslib" != yes; then
+  AC_MSG_ERROR([HTSlib development files not found
+
+ShoRAH uses HTSlib to parse bioinformatics file formats etc.
+You may need to configure --with-htslib=DIR to locate the appropriate HTSlib.
+
+FAILED.  You must supply an HTSlib in order to build ShoRAH successfully.])
+fi
```
Comment (SoapZA) [inline configure.ac]: way too complex, just use pkg-config, it works fine

File: m4/ax_with_htslib.m4 [hunk 6]
```diff
@@ -0,0 +1,154 @@
+# ===========================================================================
```
Comment (SoapZA) [inline m4/ax_with_htslib.m4]: remove this file, it's one of the awful ideas of John Marshall

File: configure.ac (hunk 4 above)
Comment (DrYak) [reply configure.ac]: Thank you for reminding me that one I've completely forgotten it...

**Comment by DrYak** [comment]
Well, basically, I was replicating the way [samtools build](https://github.com/samtools/samtools/blob/089626256f93cae0a6da89580e924f07ffc853c6/configure.ac#L33) :
 - they don't rely on `pkg-config`, but on the custom M4 script
 - therefore the slightly unusual variable names `HTSLIB_CPPFLAGS` and `HTSLIB_LDFLAGS` (which only cover the `-L` flag)

But if you don't like that approach I can convert it to [`PKG_CHECK_MODULES`](https://autotools.io/pkgconfig/pkg_check_modules.html)

**Comment by dirmeier** [comment]
David, according to you the whole world writes awful build systems :D. You should write a tutorial: "Build systems for C++, the definite guide."

**Comment by ozagordi** [comment]
Hi guys. Many thanks for working on this. I'm just observing from a distance because I'm not able to give some constructive feedback here.

**Comment by SoapZA** [comment]
@dirmeier that is technically correct, but if you use a non Turing-complete build system with proper across-language integration designed as a first-class feature, all of the pain goes away.

**Comment by ozagordi** [comment]
@SoapZA 
This reminds me that
> git gets easier once you get the basic idea that branches are homeomorphic endofunctors mapping submanifolds of a Hilbert space."

File: configure.ac (hunk 5 above)
Comment (SoapZA) [reply configure.ac]: `PKG_CHECK_MODULES([HTSLIB], [htslib])`

**Comment by SoapZA** [comment]
image

File: src/cpp/fil.cpp [hunk 7]
```diff
@@ -8,18 +26,26 @@ adapted from samtools/calDep.c
...
 #include <iostream>
 #include <map>
 #include <string>
-#include "faidx.h"
-#include "sam.h"
+#include <functional>
+#include <limits>
+
+#include <string.h>
+#include <assert.h>
```
Comment (SoapZA) [inline src/cpp/fil.cpp]: Also, this is C++, so use the `<cXXX>` instead

Comment (DrYak) [reply src/cpp/fil.cpp]: Well, in my opinion [the jury is still out](https://stackoverflow.com/a/8380835/16168) which practice is better.

Basically, `xxx.h` guarantees you that the functions will be in the global name space (and they might appear in the `std::` name space too) and `cxxx` guarantees you that the functions will be int he standart name space (And they might appear in the global one, as is the case with gnu).

Also, `cxxx` can provide multiple signatures and overload functions.
While it makes definitely sense for some functions (`std::abs` will be defined separately for floating-point types and integers, whereas `abs` might be integer-only and corrupt floating point inputs), in our case the point is moot as memory and string functions always work on the same pointer types and they will get replaced by intrinsic inlines and get optimized according to surrounding code so constness won't matter much in practice.

But okay, I'll replace the `cxxx` headers and `std::` prefixes

**Comment by DrYak** [comment]
Okay added the requested changes (and cleaned up slightly the commit log)

Also : got the data files from Susie, I'll see if the bugs (negative counts) still appear with HTSlib

**Comment by DrYak** [comment]
Small update: 
the data files that Susie gave me didn't generate any negative count.

Yippee !!!

Yippee!!!

But it's still generating zero counts
(although the previous steps does list SNVs, `fil` - the strand bias step - finds 0 of them).

That might require some further error checking.

**Comment by DrYak** [comment]
Hello, small message to ask if you did manage to find time for having a look at the requested changes I pushed 2 weeks ago ? Do they suite you?

(Or good luck with the work if you're over-whelmed with something else)

**Comment by SoapZA** [comment]
@DrYak fixed meson, should be working now. Could you please take it for a run with the test data you have (which we should be adding to Travis + something like [cram](https://pypi.org/project/cram/) tests) please?

File: src/cpp/fil.cpp [hunk 8]
```diff
@@ -52,7 +52,7 @@ static int re_init_tmp(tmpstruct_t* tmp)
 }
 
 // callback for bam_plbuf_init()
-static int pileup_func(uint32_t tid, uint32_t pos, int n, const bam_pileup1_t* pl, void* data)
+static int pileup_func(uint32_t, uint32_t pos, int n, const bam_pileup1_t* pl, void* data)
 {
     tmpstruct_t* tmp = (tmpstruct_t*)data;
     std::map<char, int> st_freq_all;
```
Comment (DrYak) [inline src/cpp/fil.cpp]: Eventually, we should switch from char to an enum :
We don't have a large amount of different states to map, only *f*orward vs *r*everse.

File: src/cpp/b2w.cpp [hunk 9]
```diff
@@ -172,14 +173,11 @@ int main(int argc, char* argv[])
...
-    auto process_ROI = [idx, ln, &header, &inFile, &fai, &reads, &covOut, &plp_iter, param] (const int tid, const int roi_b, const uint32_t roi_e) {
+    auto process_ROI = [idx, ln, &header, &inFile, &reads, &covOut, &plp_iter, param] (const int tid, const int roi_b, const uint32_t roi_e) {
         const auto lnth = ln[tid]; // this chromosome ID's length  as is defined in the header
 
         {
-            int rLen[lnth];  // list for read start coverage
-            for (unsigned j = 0; j < lnth; j++) {
-                rLen[j] = 0;
-            }
+            std::vector<int> rLen(lnth, 0);  // list for read start coverage
```
Comment (DrYak) [inline src/cpp/b2w.cpp]: :+1:

**Comment by ozagordi** [comment]
I'm trying to install, but `meson builddir` gives me

```
Native dependency htslib found: NO found '1.3.1' but need: '>=1.7'
Dependency htslib found: NO

src/cpp/meson.build:17:0: ERROR:  Dependency 'htslib' not found, tried Extra Frameworks and Pkg-Config:

Invalid version of dependency, need 'htslib' ['>=1.7'] found '1.3.1'.
```

I'm working in a conda environment where I manually installed htslib. No idea where 1.3.1 comes from, to be honest.

**Comment by DrYak** [comment]
100% agree regarding VLAs. That was on my "*should be eventually done some days*" list.
(Together with replacing the maps from char to enum, see comment above. Probably some left-over python-ism from older pure python proof of concept that got rewritten into C++ ?)

Eventually, as part of another branch, I should move the '*gathering/stats*' part of outside of the `pileup_func` call back function in fil.cpp (Should go after the `bam_plp_reset` of `main()`).

Will do tests as soon as I've got some free time (SIB Days 2018 are coming now), and will come back in touch with you.
Thanks for the meson help.

**Comment by SoapZA** [comment]
@ozagordi sounds line an ancient bioconda setup. Nowadays bioconda only uses htslib 1.7 (for long CIGAR support on PacBio and Oxford Nanopore data). We can drop it to 1.4, but 1.3.1 is definitely too low (1.3 and below have undefined behaviour issues and also break the ABI between 1.3 and 1.4).

File: src/cpp/fil.cpp (hunk 8 above)
Comment (SoapZA) [reply src/cpp/fil.cpp]: sure (`enum class` to be certain, standard C enums are smelly too)

**Comment by SoapZA** [comment]
@ozagordi also, remember, you need to set up `PKG_CONFIG_LIBDIR` to point to the directory in the bioconda prefix where `htslib.pc` (and likely `zlib.pc`) are located, otherwise Meson will find the system Htslib.

**Comment by ozagordi** [comment]
Thanks, that worked.
Now, what am I doing wrong?

```
dyld: Library not loaded: @rpath/libhts.2.dylib
  Referenced from: /private/tmp/shorah/src/shorah/bin/b2w
  Reason: image not found
```
checking with `otool`
```
otool -L /private/tmp/shorah/src/shorah/bin/b2w
/private/tmp/shorah/src/shorah/bin/b2w:
	@rpath/libhts.2.dylib (compatibility version 2.0.0, current version 1.8.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

**Comment by ozagordi** [comment]
Wow. Just wow.

```
ImportError: dlopen(/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/lib-dynload/_scproxy.cpython-36m-darwin.so, 2): Symbol not found: __cg_jpeg_resync_to_restart
  Referenced from: /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
  Expected in: /Users/ozagordi/miniconda3/envs/shorah/lib/libJPEG.dylib
 in /System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
```

File: src/cpp/fil.cpp [hunk 10]
```diff
@@ -147,70 +162,140 @@ int main(int argc, char* argv[])
...
     }
-    char chr[100], ref, var, fr1[7], fr2[7], fr3[7], p1[7], p2[7], p3[7];
-    int pos, name;
-    char str_buf[200];
-    char reg[200];
-    char filename[50];
+    std::string str_buf ("");
+
+    char* filename = NULL;
+    asprintf(&filename, "SNVs_%f.txt", tmp.sig);
```
[the remark this answers is not in the document]
Comment (DrYak) [reply src/cpp/fil.cpp]: `asprintf` is also available in BSD, and saves from hair loss (by preventing buffer overflows)
;-)

But yeah, string stream should be equivalent.

**Comment by SoapZA** [comment]
@ozagordi have you sourced the conda env?

File: src/cpp/fil.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (SoapZA) [reply src/cpp/fil.cpp]: it doesn't matter, it's not part of POSIX. What if people want to use ShoRAH on Illumos? Use portable ISO/POSIX constructs please.

[the remark this answers is not in the document]
Comment (DrYak) [reply src/cpp/fil.cpp]: Huh, yes ? [What was your question about illumos](https://illumos.org/man/3c/asprintf) again ?
:-D

On security vs standard issue, I tend to take a more extreme position: to the hell with standard, pick the secure option (at least if its reasonably widespread already, which is most often the case about sane platforms). It's burden of people with insecure platforms to come up with the necessary wrappers (here: fumbling with `snprintf`'s return size and `realloc`, or wrappers around [microsoft specific functions to get it onto windows](https://stackoverflow.com/questions/40159892/using-asprintf-on-windows)).

So i tend to happily use the auto-allocating overflow-free functions even if they aren't ISO/POSIX. Nearly all platform worth their salt have them anyway.

Put me in the same loonie bin as Theo de Raadt, if you wish. ;-)

(and stringstreams will solve the allocation problem in our case anyway)

**Comment by ozagordi** [comment]
Not working today, but yes. I remember I had activated the env. ( recent conda version requires `conda activate envname`)

**Comment by ozagordi** [comment]
So, compiling and running the new version on OSX is proving much more difficult than expected. Now I have, after changing `DYLD_LIBRARY_PATH`
```
Traceback (most recent call last):
  File "/Users/ozagordi/miniconda3/envs/shorah/bin/shorah", line 11, in <module>
    load_entry_point('ShoRAH', 'console_scripts', 'shorah')()
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/site-packages/shorah/cli.py", line 150, in main
    args.func(args)
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/site-packages/shorah/cli.py", line 75, in amplicon_run
    from shorah import amplicon
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/site-packages/shorah/amplicon.py", line 42, in <module>
    from Bio import SeqIO
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/site-packages/Bio/SeqIO/__init__.py", line 326, in <module>
    from Bio._py3k import basestring
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/site-packages/Bio/_py3k/__init__.py", line 116, in <module>
    from urllib.request import urlopen, Request, urlretrieve, urlparse, urlcleanup
  File "/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/urllib/request.py", line 2585, in <module>
    from _scproxy import _get_proxy_settings, _get_proxies
ImportError: dlopen(/Users/ozagordi/miniconda3/envs/shorah/lib/python3.6/lib-dynload/_scproxy.cpython-36m-darwin.so, 2): Library not loaded: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
  Referenced from: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/vecLib
  Reason: Incompatible library version: vecLib requires version 1.0.0 or later, but libLAPACK.dylib provides version 0.0.0
```
I'm not sure how to proceed. Is this just my problem? Just a meson problem? How does it impact the overall portability? I admit that I was trying many different things and I kind of lost track of the changes I made.

**Comment by DrYak** [comment]
Update :
- I've added the update requested by @SoapZA 
- Test passes, Valgrind is happy.

- I've done attempts at fixing the zero counts (moving the count resetting and final statistics computation out of the call back loop). Saddly it doesn't help.
- I'll clean up the code and push the commits soonish.

File: configure.ac [hunk 11]
```diff
@@ -89,6 +89,7 @@ AS_IF([test "x${enable_flags_setting}" = "xyes" && test "x${enable_debug}" = "xn
 
 	AC_LANG([C++])
 	AX_APPEND_COMPILE_FLAGS([${cflags_test}], [CXXFLAGS])
+	AX_CXX_COMPILE_STDCXX([11], [ext], [mandatory])
```
Comment (SoapZA) [inline configure.ac]: why are you enabling GNUisms?

Comment (DrYak) [reply configure.ac]: `struct` members initialization is only available in C++20,
but sadly, several environment, such as bioconda are stuck with an old GCC 4.9 that is barely C++11 (it even misses the regex from libstdc++)

enabling the GNU-ism is a work around that works on both GCC and LLVM.

**Comment by DrYak** [comment]
Hello guys,

sorry for the slow news, the heat wave managed to render my last few brain-cells inoperative, and I haven't been as efficient as I would have wanted.

Here's my latest push with the cleaned version of the last edits.

Speaking of brain not working at 100%, I missed a few things when I separated counts and statistics : actually after cleaning up it turns out it DOES solve the 'zero counts' bug, in all but a few rare corner case. (The corner case is due to how B2W and FIL behave around indels, see comments in code).

Unless @SoapZA has some cosmetic suggestions, I think this branch is done.

(The B2W and FIL case should be handled separately)

File: configure.ac (hunk 11 above)
Comment (pb-dseifert) [reply configure.ac]: what struct member initializer? what about aggregate initialization?

Comment (SoapZA) [reply configure.ac]: What `struct` initializers? Do you mean `designated initializers`? If so, either use aggregate initialization or define a constructor. Using GNUisms for that is not a win.

File: src/cpp/b2w.cpp [hunk 12]
```diff
@@ -30,138 +30,78 @@ UNSW, Sydney Australia
...
     std::ofstream covOut;  // stores window coverages
 
-    tmpstruct_t tmp;  // data for callback functions
+    paramstruct_t param = {  // data for callback functions
+        // based on default from cli.py and shotgun.py
+        .win = 201,   // window size
+        .inc = 201/3, // incr = size / shifts
+        .min_overlap = (int) std::round(201. * 0.85), // min_overlap = size * win_min_ext
+        .max = 10000 / 201 // max_c = max_coverage / size
+    };
```
Comment (SoapZA) [inline src/cpp/b2w.cpp]: all of this can be done with simple aggregate initialization

Comment (SoapZA) [reply src/cpp/b2w.cpp]: indeed, this is the only line with designated initializers. Enabling GNUisms over ISO C++ for one initialization will not fly.

**Comment by DrYak** [comment]
Thanks for the fixes ! :+1:

**Pull request merged**